src/chrome/internal/sources/sourceTextRetriever.ts

```diff
@@ -20,47 +20,91 @@ let localize = nls.loadMessageBundle();
-                return utils.readFileP(loadedSource.identifier.textRepresentation);
-            } else {
-                // We'll need to figure out what is the right thing to do for SourceScriptRelationship.Unknown
-                throw new LocalizedError('error.sourceText.multipleScriptsNotSupported', localize('error.sourceText.multipleScriptsNotSupported', "Support for getting the text from dynamic sources that have multiple scripts embedded hasn't been implemented yet"));
-            }
+            const sourceRetrievability = this._retrievability(this._scriptSources, loadedSource);
+            text = sourceRetrievability.retrieve();
             this._sourceToText.set(loadedSource, text);
         }
 
         return text;
     }
 
+    public retrievability(source: ILoadedSource): IPossiblyRetrievableText {
+        return this._retrievability(this._scriptSources, source);
+    }
+
     public toString(): string {
         return `Sources text logic\n${printIterable('sources in cache', this._sourceToText.keys())}`;
     }
-}
\ No newline at end of file
+}
+
+export type GetSourceTextRetrievability = (scriptSources: IScriptSourcesRetriever, loadedSource: ILoadedSource) => IPossiblyRetrievableText;
+
+// We use the retrievability instead of the text method to figure out if a loadedSource that we'll send to the client
+// is retrievable, and thus, if it should include a sourceReference in it or not
+export function getSourceTextRetrievability(scriptSources: IScriptSourcesRetriever, loadedSource: ILoadedSource): IPossiblyRetrievableText {
+    const scripts = loadedSource.scriptMapper().scripts;
+    if (loadedSource.sourceScriptRelationship === SourceScriptRelationship.SourceIsSingleScript && scripts.length === 1) {
+        const singleScript = singleElementOfArray(scripts);
+        return new RetrievableText(() => scriptSources.getScriptSource(singleScript));
+    } else if (loadedSource.sourceScriptRelationship === SourceScriptRelationship.SourceIsSingleScript && scripts.length >= 2) {
+        /**
+         * We have two or more scripts associated with this source. At the moment we don't have any further support for this scenario
+         * so we just return the source of the first script. This won't be ideal if for some reason the source of both scripts
+         * isn't the same.
+         */
+        logger.warn(`${loadedSource} is associated with several ${printArray('scripts', scripts)} returning arbitrarily the source of the first one`);
+        const singleScript = scripts[0];
+        return new RetrievableText(() => scriptSources.getScriptSource(singleScript));
+    } else if (loadedSource.contentsLocation === ContentsLocation.PersistentStorage) {
+        // If this is a file, we don't want to cache it, so we return the contents immediately
+        return new RetrievableText(() => utils.readFileP(loadedSource.identifier.textRepresentation));
```

Comment:
When does the debug adapter have to return the text of a file that's on disk? Shouldn't vscode/VS always be able to read it?

Reply:
As far as I know that's accurate. I implemented this because the protocol interface SourceArguments says:
   * Specifies the source content to load. Either source.path or source.sourceReference must be specified.
The source.path part made me think that someone might decide to do this with a path...

Reply:
Hm, I guess we technically need to handle it then.